# meta-xtrome-analysis for RNASeq datasets

**Folders containts**:<br>
**notebooks**<br>
Phython 3+ scritps <br>
**Data_acc**<br>
SRA/NCBI SEARCHING TRANSCRIPTOMES USING THE SRA RUN SELECTOR FILTER<br>
**R_code**<br>
Code files to perform the WGCNA Standar analysis <br>
**Uniprot_defense_data**<br>
Code to filter several UNIPROT Taxonomies <br>
**meta-data**<br>
Files to perform other processes like:
Araport11_functional_descriptions <br>
Athal_Traits2.tsv
Athal_Traits_Feb2021.tsv
athal_GO_terms.txt

**WGCNA Coexpression Ntw(s) were built with the help of online public resources. Here I share some of them:**<br>
Please be aware you need to adjust code to your own needs <br>
https://horvath.genetics.ucla.edu/html/CoexpressionNetwork/Rpackages/WGCNA/ <br>
http://pklab.med.harvard.edu/scw2014/WGCNA.html <br>
https://wikis.utexas.edu/display/bioiteam/Clustering+using+WGCNA <br>
https://www.polarmicrobes.org/weighted-gene-correlation-network-analysis-wgcna-applied-to-microbial-communities/ <br>





