{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "# Getting gene lenghts from GTF/GFF file.\n",
    "\n",
    "Made by: Cynthia Soto <br>\n",
    "Date: March 12nd 2021 <br>\n",
    "Last md: March 17th <br>\n",
    "\n",
    "**Assumptions:**<br>\n",
    "You have a GTF file to extract the get lengths (GTF or GFF file)  <br>"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Libraries required\n",
    "import os \n",
    "#import glob\n",
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Directory changed: /home/cyntsc/Proyectos/tesis-code/meta-xtrome-analysis/meta-data\n"
     ]
    }
   ],
   "source": [
    "try:\n",
    "    # Change the current working Directory      \n",
    "    os.chdir(\"/home/cyntsc/Proyectos/tesis-code/meta-xtrome-analysis/meta-data/\")\n",
    "    print(\"Directory changed:\", os.getcwd() )\n",
    "except OSError:\n",
    "    print(\"Can't change the Current Working Directory\")    \n",
    "s_path = str(os.getcwd())+'/*_Log2.csv'       \n",
    "#print(os.listdir(os.getcwd())       "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "**Araport11_GFF3_genes_transposons.201606.gtf** es la ultima anotación liberada para el genoma **TAIR10** en Julio del 2016. En esta anotación se añadio información para 253 genes codificantes, informacion de mRNAs y lncRNAs, entre otras caracteristicas genómicas <br>\n",
    "La estimación de los valores de expresión fue realizada con **HTSeq** por lo que se utilizó una versión simplificada del archivo **gtf** llamada **gff** para cumplir con los requisitos de la herramienta de cuantificación (descargada de phytozone)<br> \n",
    "Para normalizar los datos y hacerlos comparables entre sí aplicaré la **normalización TPM**. La cual se calcula como sigue:\n",
    "![TPM](figuras_notebook/TPM.gif)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Para aplicar esta normalización se requiere la **longitud del gen**, por lo que voy a filtrar las columnas 3,4,5 y 8 para extraer esta información, donde la longitud estará definida por la diferencia entre las columnas 4 y 5. <br>\n",
    "Nota que esto dará lugar a una generalización sobre las longitudes de los genes que no es necesariamente precisa, pero que sirve para hacer un primer análisis exploratorio de los datos. Existen otras formas de alcanzar este meta y todos los métodos existentes hoy en día presentan pros y contras:\n",
    "https://www.biostars.org/p/91218/#91233; https://www.biostars.org/p/112735/; https://www.biostars.org/p/294545/#294591 <br>\n",
    "![gene_annot_example](figuras_notebook/gene_annot_file.png)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>transcript</td>\n",
       "      <td>3631</td>\n",
       "      <td>5899</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>exon</td>\n",
       "      <td>3631</td>\n",
       "      <td>3913</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>exon</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>exon</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>exon</td>\n",
       "      <td>4706</td>\n",
       "      <td>5095</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668992</th>\n",
       "      <td>exon</td>\n",
       "      <td>363725</td>\n",
       "      <td>364042</td>\n",
       "      <td>transcript_id \"ATMG01400.1\"; gene_id \"ATMG01400\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668993</th>\n",
       "      <td>CDS</td>\n",
       "      <td>363725</td>\n",
       "      <td>364042</td>\n",
       "      <td>transcript_id \"ATMG01400.1\"; gene_id \"ATMG01400\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668994</th>\n",
       "      <td>transcript</td>\n",
       "      <td>366086</td>\n",
       "      <td>366700</td>\n",
       "      <td>transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668995</th>\n",
       "      <td>exon</td>\n",
       "      <td>366086</td>\n",
       "      <td>366700</td>\n",
       "      <td>transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>668996</th>\n",
       "      <td>CDS</td>\n",
       "      <td>366086</td>\n",
       "      <td>366700</td>\n",
       "      <td>transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>668997 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                 2       3       4  \\\n",
       "0       transcript    3631    5899   \n",
       "1             exon    3631    3913   \n",
       "2             exon    3996    4276   \n",
       "3             exon    4486    4605   \n",
       "4             exon    4706    5095   \n",
       "...            ...     ...     ...   \n",
       "668992        exon  363725  364042   \n",
       "668993         CDS  363725  364042   \n",
       "668994  transcript  366086  366700   \n",
       "668995        exon  366086  366700   \n",
       "668996         CDS  366086  366700   \n",
       "\n",
       "                                                        8  \n",
       "0       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "1       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "2       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "3       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "4       transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";  \n",
       "...                                                   ...  \n",
       "668992  transcript_id \"ATMG01400.1\"; gene_id \"ATMG01400\";  \n",
       "668993  transcript_id \"ATMG01400.1\"; gene_id \"ATMG01400\";  \n",
       "668994  transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";  \n",
       "668995  transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";  \n",
       "668996  transcript_id \"ATMG01410.1\"; gene_id \"ATMG01410\";  \n",
       "\n",
       "[668997 rows x 4 columns]"
      ]
     },
     "execution_count": 47,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df=pd.read_csv(\"Araport11_GFF3_genes_transposons.201606.gtf\", header=None, usecols=[2,3,4,8], sep='\\t')    ## ,nrows=100  usecols=[1,3] / Araport11_GFF3.gff\n",
    "df\n",
    "## el valor de la columna 2 contiene las caracteristicas (feature) genómicas. \n",
    "## el valor \"CDS\" contiene la longitud máxima del gen, que deberá ser calculada con: columna 4 - columna 3"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "## filtering by column 2 which contains the available features\n",
    "f_genes = df[2]=='CDS' # \n",
    "df1 = df[f_genes]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "(286355, 4)\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>2</th>\n",
       "      <th>3</th>\n",
       "      <th>4</th>\n",
       "      <th>8</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4706</td>\n",
       "      <td>5095</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>5174</td>\n",
       "      <td>5326</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>CDS</td>\n",
       "      <td>5439</td>\n",
       "      <td>5630</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>CDS</td>\n",
       "      <td>7315</td>\n",
       "      <td>7450</td>\n",
       "      <td>transcript_id \"AT1G01020.2\"; gene_id \"AT1G01020\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      2     3     4                                                  8\n",
       "7   CDS  3760  3913  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "8   CDS  3996  4276  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "9   CDS  4486  4605  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "10  CDS  4706  5095  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "11  CDS  5174  5326  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "12  CDS  5439  5630  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "22  CDS  7315  7450  transcript_id \"AT1G01020.2\"; gene_id \"AT1G01020\";"
      ]
     },
     "execution_count": 49,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "print(df1.shape)\n",
    "df1.head(7)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Description</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4706</td>\n",
       "      <td>5095</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>5174</td>\n",
       "      <td>5326</td>\n",
       "      <td>transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Feature  Start   End                                        Description\n",
       "7      CDS   3760  3913  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "8      CDS   3996  4276  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "9      CDS   4486  4605  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "10     CDS   4706  5095  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";\n",
       "11     CDS   5174  5326  transcript_id \"AT1G01010.1\"; gene_id \"AT1G01010\";"
      ]
     },
     "execution_count": 50,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1.columns = ['Feature','Start','End','Description']\n",
    "df1.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>transcript_id \"AT1G01010.1\";</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>transcript_id \"AT1G01010.1\";</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>transcript_id \"AT1G01010.1\";</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                              0            1\n",
       "7  transcript_id \"AT1G01010.1\";  AT1G01010\";\n",
       "8  transcript_id \"AT1G01010.1\";  AT1G01010\";\n",
       "9  transcript_id \"AT1G01010.1\";  AT1G01010\";"
      ]
     },
     "execution_count": 51,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## split the 'Description' field to get the gene_id alone\n",
    "df_gene = df1[\"Description\"].str.split(' gene_id \"', n = 1, expand = True) #,  expand = True\n",
    "df_gene.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cyntsc/Proyectos/tesis-code/jn_tesis_env/lib/python3.6/site-packages/ipykernel_launcher.py:2: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \n",
      "/home/cyntsc/Proyectos/tesis-code/jn_tesis_env/lib/python3.6/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Name_gene</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>AT1G01010\";</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  Feature  Start   End    Name_gene\n",
       "7     CDS   3760  3913  AT1G01010\";\n",
       "8     CDS   3996  4276  AT1G01010\";\n",
       "9     CDS   4486  4605  AT1G01010\";"
      ]
     },
     "execution_count": 52,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# making separate first name column from new data frame \n",
    "df1[\"Name_gene\"] = df_gene[1]\n",
    "df1.drop(columns = ['Description'], inplace = True) \n",
    "df1.head(3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>0</th>\n",
       "      <th>1</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>;</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>;</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "            0  1\n",
       "7   AT1G01010  ;\n",
       "8   AT1G01010  ;\n",
       "9   AT1G01010  ;\n",
       "10  AT1G01010  ;\n",
       "11  AT1G01010  ;"
      ]
     },
     "execution_count": 53,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene = df1[\"Name_gene\"].str.split('\"', n = 1, expand = True) \n",
    "df_gene.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "/home/cyntsc/Proyectos/tesis-code/jn_tesis_env/lib/python3.6/site-packages/ipykernel_launcher.py:1: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  \"\"\"Entry point for launching an IPython kernel.\n",
      "/home/cyntsc/Proyectos/tesis-code/jn_tesis_env/lib/python3.6/site-packages/pandas/core/frame.py:3997: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  errors=errors,\n"
     ]
    },
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Feature</th>\n",
       "      <th>Start</th>\n",
       "      <th>End</th>\n",
       "      <th>Genes</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3760</td>\n",
       "      <td>3913</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>CDS</td>\n",
       "      <td>3996</td>\n",
       "      <td>4276</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4486</td>\n",
       "      <td>4605</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>CDS</td>\n",
       "      <td>4706</td>\n",
       "      <td>5095</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>CDS</td>\n",
       "      <td>5174</td>\n",
       "      <td>5326</td>\n",
       "      <td>AT1G01010</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "   Feature  Start   End      Genes\n",
       "7      CDS   3760  3913  AT1G01010\n",
       "8      CDS   3996  4276  AT1G01010\n",
       "9      CDS   4486  4605  AT1G01010\n",
       "10     CDS   4706  5095  AT1G01010\n",
       "11     CDS   5174  5326  AT1G01010"
      ]
     },
     "execution_count": 54,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df1['Genes']=df_gene[0]\n",
    "df1.drop(columns = ['Name_gene'], inplace = True) \n",
    "df1.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Jump to the aggregation functions to do it faster\n",
    "Or do it by step by step"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "metadata": {},
   "outputs": [],
   "source": [
    "#df_gene_len = df1.groupby(['Genes','Feature','Start','End'])[\"ID\"].count()\n",
    "#df_gene_len.head(20)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 56,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Genes\n",
       "AT1G01010     6\n",
       "AT1G01020    42\n",
       "AT1G01030     3\n",
       "AT1G01040    40\n",
       "AT1G01050    16\n",
       "AT1G01060    56\n",
       "AT1G01070    11\n",
       "AT1G01080    11\n",
       "AT1G01090     3\n",
       "AT1G01100    13\n",
       "Name: Genes, dtype: int64"
      ]
     },
     "execution_count": 56,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## grouping by ID to know the distributions based on the length\n",
    "s_by_gene = df1.groupby(\"Genes\")[\"Genes\"].count()\n",
    "s_by_gene.head(10)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 57,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27655"
      ]
     },
     "execution_count": 57,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "## checking if number of features is correct\n",
    "#print(s_by_gene.dtypes)\n",
    "len(s_by_gene)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 58,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Genes\n",
      "ATMG01410    1\n",
      "AT2G16570    1\n",
      "AT5G07545    1\n",
      "AT2G16500    1\n",
      "AT2G16450    1\n",
      "AT5G07572    1\n",
      "AT5G07580    1\n",
      "AT5G07610    1\n",
      "AT2G16280    1\n",
      "AT2G16220    1\n",
      "Name: Genes, dtype: int64\n",
      "Genes\n",
      "AT5G27970    222\n",
      "AT3G57410    224\n",
      "AT3G08840    234\n",
      "AT1G20060    235\n",
      "AT5G24740    271\n",
      "AT4G15215    273\n",
      "AT5G06120    283\n",
      "AT1G24706    288\n",
      "AT2G33240    324\n",
      "AT1G48090    381\n",
      "Name: Genes, dtype: int64\n"
     ]
    }
   ],
   "source": [
    "## a taste of information about data\n",
    "s_by_gene.sort_values(inplace=True)\n",
    "print(s_by_gene.head(10))\n",
    "print(s_by_gene.tail(10))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "381\n",
      "1\n"
     ]
    }
   ],
   "source": [
    "print(s_by_gene.max())\n",
    "print(s_by_gene.min())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAD4CAYAAAA6j0u4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAYF0lEQVR4nO3dfbRddX3n8fdHkKf6AEhKWQlMokYtWrXxCnTZOg4MENAaOkNduJwh47DMTMWOD52loF2No2UtnZkaZUaxVFLBWgFRS8ZiMQKtM2sND0GQR4FbUEkKJjU8aLEw0e/8cX4XTuO9yUnu3eecXN6vtc66e//2b5/9PRtvPu7f+d29U1VIktSlZ4y6AEnS/GfYSJI6Z9hIkjpn2EiSOmfYSJI6t/eoCxi2Qw45pBYvXjzqMiRpj3LjjTf+fVUt2N39n3Zhs3jxYjZs2DDqMiRpj5Lke7PZ32E0SVLnDBtJUuc6C5ska5NsTnLbNNt+L0klOaStJ8m5SSaT3JJkWV/flUnuaa+Vfe2vSnJr2+fcJOnqs0iSZqfLK5vPAsu3b0xyOHAC8P2+5pOApe21Cjiv9T0YWA0cDRwFrE5yUNvnPOBtffv93LEkSeOhs7Cpqm8CW6fZtAZ4L9B/U7YVwEXVcy1wYJLDgBOB9VW1taoeAtYDy9u251TVtdW7udtFwCldfRZJ0uwM9TubJCuATVX17e02LQTu71vf2Np21L5xmnZJ0hga2tTnJAcA76c3hDZUSVbRG57jiCOOGPbhJelpb5hXNi8AlgDfTvJdYBHwrSS/BGwCDu/ru6i17ah90TTt06qq86tqoqomFizY7b9JkiTtpqGFTVXdWlW/WFWLq2oxvaGvZVX1ILAOOL3NSjsGeKSqHgCuBE5IclCbGHACcGXb9miSY9ostNOBy4f1WSRJu6azYbQkXwBeBxySZCOwuqoumKH7FcDJwCTwGPBWgKramuTDwA2t34eqamrSwdvpzXjbH/hae3Vqzfq7n1x+9/Ev6vpwkjRvdBY2VfXmnWxf3LdcwJkz9FsLrJ2mfQPwstlVKUkaBu8gIEnqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqnGEjSeqcYSNJ6pxhI0nqXGdhk2Rtks1Jbutr+29JvpPkliRfSXJg37azk0wmuSvJiX3ty1vbZJKz+tqXJLmutV+SZJ+uPoskaXa6vLL5LLB8u7b1wMuq6uXA3cDZAEmOBE4DXtr2+VSSvZLsBXwSOAk4Enhz6wvwUWBNVb0QeAg4o8PPIkmahc7Cpqq+CWzdru3rVbWtrV4LLGrLK4CLq+rxqroPmASOaq/Jqrq3qp4ALgZWJAlwLHBZ2/9C4JSuPoskaXZG+Z3Nvwe+1pYXAvf3bdvY2mZqfx7wcF9wTbVPK8mqJBuSbNiyZcsclS9JGtRIwibJB4BtwOeHcbyqOr+qJqpqYsGCBcM4pCSpz97DPmCSfwe8ATiuqqo1bwIO7+u2qLUxQ/sPgQOT7N2ubvr7S5LGzFCvbJIsB94LvLGqHuvbtA44Lcm+SZYAS4HrgRuApW3m2T70JhGsayF1DXBq238lcPmwPockadd0OfX5C8D/BV6cZGOSM4D/CTwbWJ/k5iSfBqiq24FLgTuAvwLOrKqftquWdwBXAncCl7a+AO8D3pNkkt53OBd09VkkSbPT2TBaVb15muYZA6GqzgHOmab9CuCKadrvpTdbTZI05ryDgCSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOdhU2StUk2J7mtr+3gJOuT3NN+HtTak+TcJJNJbkmyrG+fla3/PUlW9rW/KsmtbZ9zk6SrzyJJmp0ur2w+Cyzfru0s4KqqWgpc1dYBTgKWttcq4DzohROwGjgaOApYPRVQrc/b+vbb/liSpDHRWdhU1TeBrds1rwAubMsXAqf0tV9UPdcCByY5DDgRWF9VW6vqIWA9sLxte05VXVtVBVzU916SpDEz7O9sDq2qB9ryg8ChbXkhcH9fv42tbUftG6dpn1aSVUk2JNmwZcuW2X0CSdIuG9kEgXZFUkM61vlVNVFVEwsWLBjGISVJfYYdNj9oQ2C0n5tb+ybg8L5+i1rbjtoXTdMuSRpDww6bdcDUjLKVwOV97ae3WWnHAI+04bYrgROSHNQmBpwAXNm2PZrkmDYL7fS+95IkjZm9u3rjJF8AXgcckmQjvVllHwEuTXIG8D3gTa37FcDJwCTwGPBWgKramuTDwA2t34eqamrSwdvpzXjbH/hae0mSxlBnYVNVb55h03HT9C3gzBneZy2wdpr2DcDLZlOjJGk4vIOAJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpcwOFTZJf6boQSdL8NeiVzaeSXJ/k7Ume22lFkqR5Z6CwqarfAN5C76mZNyb58yTHd1qZJGneGPg7m6q6B/h94H3APwfOTfKdJP+qq+IkSfPDoN/ZvDzJGuBO4FjgN6vql9vymg7rkyTNA4M+qfN/AJ8B3l9VP5lqrKq/S/L7nVQmSZo3Bg2b1wM/qaqfAiR5BrBfVT1WVZ/rrDpJ0rww6Hc23wD271s/oLVJkrRTg4bNflX146mVtnxANyVJkuabQcPmH5Ism1pJ8irgJzvov0NJ3p3k9iS3JflCkv2SLElyXZLJJJck2af13betT7bti/ve5+zWfleSE3e3HklStwYNm3cBX0zyv5P8H+AS4B27c8AkC4H/BExU1cuAvYDTgI8Ca6rqhcBDwBltlzOAh1r7mtaPJEe2/V4KLKf3h6d77U5NkqRuDfpHnTcALwF+B/iPwC9X1Y2zOO7ewP5J9qY3HPcAvWnUl7XtFwKntOUVbZ22/bgkae0XV9XjVXUfMAkcNYuaJEkdGXQ2GsCrgcVtn2VJqKqLdvWAVbUpyX8Hvk9vKO7rwI3Aw1W1rXXbCCxsywuB+9u+25I8AjyvtV/b99b9+/wTSVYBqwCOOOKIXS1ZkjRLA4VNks8BLwBuBn7amgvY5bBJchC9q5IlwMPAF+kNg3Wmqs4HzgeYmJioLo8lSfp5g17ZTABHVtVc/EP9L4H7qmoLQJIvA68BDkyyd7u6WQRsav030bsn28Y27PZc4Id97VP695EkjZFBJwjcBvzSHB3z+8AxSQ5o370cB9wBXAOc2vqsBC5vy+vaOm371S301gGntdlqS4ClwPVzVKMkaQ4NemVzCHBHkuuBx6caq+qNu3rAqrouyWXAt4BtwE30hrj+Erg4yR+2tgvaLhcAn0syCWylNwONqro9yaX0gmobcObUHQ4kSeNl0LD54FwetKpWA6u3a76XaWaTVdU/Ar89w/ucA5wzl7VJkubeQGFTVX+T5J8BS6vqG0kOoPf3MZIk7dSgjxh4G72/cfnj1rQQ+IuOapIkzTODThA4k96MsUfhyQep/WJXRUmS5pdBw+bxqnpiaqVNQfbvVSRJAxk0bP4myfvp3WLmeHp/iPm/uitLkjSfDBo2ZwFbgFuB/wBcAfiETknSQAadjfYz4E/aS5KkXTLovdHuY5rvaKrq+XNekSRp3tmVe6NN2Y/eH1kePPflSJLmo0GfZ/PDvtemqvo48PpuS5MkzReDDqMt61t9Br0rnV15Fo4k6Wls0MD4o77lbcB3gTfNeTWSpHlp0Nlo/6LrQiRJ89egw2jv2dH2qvrY3JQjSZqPdmU22qvpPbAM4DfpPajsni6KkiTNL4OGzSJgWVX9CCDJB4G/rKp/01VhkqT5Y9Db1RwKPNG3/kRrkyRppwa9srkIuD7JV9r6KcCFnVQkSZp3Bp2Ndk6SrwG/0ZreWlU3dVeWJGk+GXQYDeAA4NGq+gSwMcmSjmqSJM0zgz4WejXwPuDs1vRM4M+6KkqSNL8MemXzW8AbgX8AqKq/A569uwdNcmCSy5J8J8mdSX4tycFJ1ie5p/08qPVNknOTTCa5pf/WOUlWtv73JFm5u/VIkro1aNg8UVVFe8xAkl+Y5XE/AfxVVb0EeAVwJ70HtF1VVUuBq9o6wEnA0vZaBZzXajgYWA0cDRwFrJ4KKEnSeBk0bC5N8sfAgUneBnyD3XyQWpLnAq8FLgCoqieq6mFgBU/NcLuQ3ow3WvtF1XNtq+Ew4ERgfVVtraqHgPXA8t2pSZLUrZ3ORksS4BLgJcCjwIuBP6iq9bt5zCX0HjH9p0leAdwIvBM4tKoeaH0e5Km/41kI3N+3/8bWNlP7dJ9hFb2rIo444ojdLFuStLt2GjZVVUmuqKpfoXf1MBfHXAb8blVdl+QTPDVk1n/Mn3sy6O6qqvOB8wEmJibm7H0lSYMZdBjtW0lePUfH3AhsrKrr2vpl9MLnB214jPZzc9u+CTi8b/9FrW2mdknSmBk0bI4Grk3yt21G2K1JbtmdA1bVg8D9SV7cmo4D7qB3k8+pGWUrgcvb8jrg9DYr7RjgkTbcdiVwQpKD2sSAE1qbJGnM7HAYLckRVfV9el/Gz6XfBT6fZB/gXuCt9ILv0iRnAN/jqYezXQGcDEwCj7W+VNXWJB8Gbmj9PlRVW+e4TknSHNjZdzZ/Qe9uz99L8qWq+tdzcdCqupneYwu2d9w0fQs4c4b3WQusnYuaJEnd2dkwWvqWn99lIZKk+WtnYVMzLEuSNLCdDaO9Ismj9K5w9m/LtPWqqud0Wp0kaV7YYdhU1V7DKkSSNH/tyiMGJEnaLYaNJKlzho0kqXOGjSSpc4aNJKlzho0kqXOGjSSpc4aNJKlzho0kqXM7fVKnprdm/d1PLr/7+BeNsBJJGn9e2UiSOmfYSJI6Z9hIkjpn2EiSOmfYSJI6Z9hIkjpn2EiSOjeysEmyV5Kbkny1rS9Jcl2SySSXJNmnte/b1ifb9sV973F2a78ryYkj+iiSpJ0Y5ZXNO4E7+9Y/CqypqhcCDwFntPYzgIda+5rWjyRHAqcBLwWWA59K4mOsJWkMjSRskiwCXg98pq0HOBa4rHW5EDilLa9o67Ttx7X+K4CLq+rxqroPmASOGsoHkCTtklFd2XwceC/ws7b+PODhqtrW1jcCC9vyQuB+gLb9kdb/yfZp9vknkqxKsiHJhi1btszhx5AkDWLoYZPkDcDmqrpxWMesqvOraqKqJhYsWDCsw0qSmlHciPM1wBuTnAzsBzwH+ARwYJK929XLImBT678JOBzYmGRv4LnAD/vap/TvI0kaI0O/sqmqs6tqUVUtpvcF/9VV9RbgGuDU1m0lcHlbXtfWaduvrqpq7ae12WpLgKXA9UP6GJKkXTBOjxh4H3Bxkj8EbgIuaO0XAJ9LMglspRdQVNXtSS4F7gC2AWdW1U+HX7YkaWdGGjZV9dfAX7fle5lmNllV/SPw2zPsfw5wTncVSpLmgncQkCR1zrCRJHXOsJEkdc6wkSR1zrCRJHXOsJEkdc6wkSR1zrCRJHXOsJEkdc6wkSR1zrCRJHXOsJEkdc6wkSR1zrCRJHVunJ5ns8das/7uJ5ffffyLRliJJI0nr2wkSZ0zbCRJnTNsJEmdM2wkSZ0zbCRJnTNsJEmdG3rYJDk8yTVJ7khye5J3tvaDk6xPck/7eVBrT5Jzk0wmuSXJsr73Wtn635Nk5bA/iyRpMKO4stkG/F5VHQkcA5yZ5EjgLOCqqloKXNXWAU4ClrbXKuA86IUTsBo4GjgKWD0VUJKk8TL0sKmqB6rqW235R8CdwEJgBXBh63YhcEpbXgFcVD3XAgcmOQw4EVhfVVur6iFgPbB8eJ9EkjSokX5nk2Qx8KvAdcChVfVA2/QgcGhbXgjc37fbxtY2U/t0x1mVZEOSDVu2bJm7DyBJGsjIwibJs4AvAe+qqkf7t1VVATVXx6qq86tqoqomFixYMFdvK0ka0EjCJskz6QXN56vqy635B214jPZzc2vfBBzet/ui1jZTuyRpzIxiNlqAC4A7q+pjfZvWAVMzylYCl/e1n95mpR0DPNKG264ETkhyUJsYcEJrkySNmVHc9fk1wL8Fbk1yc2t7P/AR4NIkZwDfA97Utl0BnAxMAo8BbwWoqq1JPgzc0Pp9qKq2DuUTSJJ2SXpfjzx9TExM1IYNG3Zr3/5HCQzCxw1Imi+S3FhVE7u7v3cQkCR1zrCRJHXOsJEkdc6wkSR1zrCRJHXOsJEkdc6wkSR1zrCRJHXOsJEkdc6wkSR1zrCRJHVuFDfifNqY6V5q3jNN0tONVzaSpM4ZNpKkzjmMNgL9w2sOqUl6OvDKRpLUOcNGktQ5h9FGzCE1SU8Hhs0YMXgkzVcOo0mSOueVzZjyD0IlzSd7fNgkWQ58AtgL+ExVfWTEJXXKEJK0J9qjwybJXsAngeOBjcANSdZV1R2jrWz4tg8hw0fSONmjwwY4CpisqnsBklwMrACedmGzvZmugLrQH2xeeUmazp4eNguB+/vWNwJHb98pySpgVVv9cZK7duNYhwB/vxv7DcvI6nvPYH08f7NjfbNjfbP34tnsvKeHzUCq6nzg/Nm8R5INVTUxRyXNOeubHeubHeubnXGvD3o1zmb/PX3q8ybg8L71Ra1NkjRG9vSwuQFYmmRJkn2A04B1I65JkrSdPXoYraq2JXkHcCW9qc9rq+r2jg43q2G4IbC+2bG+2bG+2Rn3+mC2X0VU1VwVIknStPb0YTRJ0h7AsJEkdc6wGUCS5UnuSjKZ5KxR1wOQ5LtJbk1y89SUxCQHJ1mf5J7286Ah1rM2yeYkt/W1TVtPes5t5/OWJMtGVN8Hk2xq5/DmJCf3bTu71XdXkhOHUN/hSa5JckeS25O8s7WPxTncQX1jcQ6T7Jfk+iTfbvX9l9a+JMl1rY5L2kQikuzb1ifb9sUjqu+zSe7rO3+vbO1D/x1px90ryU1JvtrW5+78VZWvHbzoTTz4W+D5wD7At4Ejx6Cu7wKHbNf2X4Gz2vJZwEeHWM9rgWXAbTurBzgZ+BoQ4BjguhHV90HgP0/T98j233lfYEn7779Xx/UdBixry88G7m51jMU53EF9Y3EO23l4Vlt+JnBdOy+XAqe19k8Dv9OW3w58ui2fBlzS8fmbqb7PAqdO03/ovyPtuO8B/hz4alufs/Pnlc3OPXlLnKp6Api6Jc44WgFc2JYvBE4Z1oGr6pvA1gHrWQFcVD3XAgcmOWwE9c1kBXBxVT1eVfcBk/T+d9CZqnqgqr7Vln8E3EnvDhljcQ53UN9MhnoO23n4cVt9ZnsVcCxwWWvf/vxNndfLgOOSZAT1zWTovyNJFgGvBz7T1sMcnj/DZuemuyXOjn7JhqWArye5Mb3b8QAcWlUPtOUHgUNHU9qTZqpnnM7pO9owxdq+YceR1teGJH6V3v/7HbtzuF19MCbnsA0B3QxsBtbTu5p6uKq2TVPDk/W17Y8AzxtmfVU1df7OaedvTZJ9t69vmtq78nHgvcDP2vrzmMPzZ9jsuX69qpYBJwFnJnlt/8bqXd+Ozbz2caunOQ94AfBK4AHgj0ZaDZDkWcCXgHdV1aP928bhHE5T39icw6r6aVW9kt6dRI4CXjKqWqazfX1JXgacTa/OVwMHA+8bRW1J3gBsrqobuzqGYbNzY3lLnKra1H5uBr5C75frB1OX2u3n5tFVCDuoZyzOaVX9oP0D8DPgT3hqmGck9SV5Jr1/yD9fVV9uzWNzDqerb9zOYavpYeAa4NfoDT9N/fF6fw1P1te2Pxf44ZDrW96GJ6uqHgf+lNGdv9cAb0zyXXpfFRxL7zlhc3b+DJudG7tb4iT5hSTPnloGTgBua3WtbN1WApePpsInzVTPOuD0NuPmGOCRvqGiodluDPy36J3DqfpOazNulgBLges7riXABcCdVfWxvk1jcQ5nqm9czmGSBUkObMv703vG1Z30/lE/tXXb/vxNnddTgavbleMw6/tO3/+RCL3vQ/rP39D++1bV2VW1qKoW0/s37uqqegtzef66nt0wH170ZobcTW8M+ANjUM/z6c30+TZw+1RN9MZMrwLuAb4BHDzEmr5Abxjl/9Eb2z1jpnrozbD5ZDuftwITI6rvc+34t7RfnsP6+n+g1XcXcNIQ6vt1ekNktwA3t9fJ43IOd1DfWJxD4OXATa2O24A/6PtduZ7eBIUvAvu29v3a+mTb/vwR1Xd1O3+3AX/GUzPWhv470lfr63hqNtqcnT9vVyNJ6pzDaJKkzhk2kqTOGTaSpM4ZNpKkzhk2kqTOGTaSpM4ZNpKkzv1/XUgXJ8jJOa0AAAAASUVORK5CYII=\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "#ax = s_by_gene.plot.kde(bw_method=1)   #bw_method=0.5\n",
    "#ax = s_by_gene.plot.box()\n",
    "ax = s_by_gene.plot.hist(bins=100, alpha=0.5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<AxesSubplot:ylabel='Frequency'>"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    },
    {
     "data": {
      "image/png": "iVBORw0KGgoAAAANSUhEUgAAAZsAAAD4CAYAAAA6j0u4AAAAOXRFWHRTb2Z0d2FyZQBNYXRwbG90bGliIHZlcnNpb24zLjMuMCwgaHR0cHM6Ly9tYXRwbG90bGliLm9yZy86wFpkAAAACXBIWXMAAAsTAAALEwEAmpwYAAAVUUlEQVR4nO3df7DddX3n8efL8MsfWEBSliHsBttMLbUWY0B2tF2rIwTYNrjLOjhtybiM6VaY0Wl31mA7hdUyozurtHSUFmsWsCrgr5KtcWlEpq5/8CMo8lNKFnFJiiQlSLA6sOh7/zifC2fivTcnufdz7sn1+Zg5c7/n/f1+z/d9PuHeF+d7Pud7UlVIktTTCxa6AUnS4mfYSJK6M2wkSd0ZNpKk7gwbSVJ3By10A+N29NFH1/Llyxe6DUk6oNxxxx3/VFVL93f/n7qwWb58OVu2bFnoNiTpgJLkO3PZ39NokqTuDBtJUneGjSSpO8NGktSdYSNJ6s6wkSR1Z9hIkrozbCRJ3Rk2kqTufuquIDAXy9d/cUGO+/AHzlqQ40rSfPGVjSSpO8NGktSdYSNJ6s6wkSR1Z9hIkrrrFjZJjk9yc5L7ktyb5F2tfkmS7UnubLczh/a5KMnWJA8kOX2ovrrVtiZZP1Q/IcmtrX5dkkN6PR9J0v7r+crmWeAPqupE4FTggiQntnWXVdVJ7bYJoK07F/glYDXw0SRLkiwBPgKcAZwIvG3ocT7YHuvngSeA8zs+H0nSfuoWNlX1aFV9vS0/BdwPHDfLLmuAa6vq6ar6NrAVOKXdtlbVQ1X1DHAtsCZJgDcCn237Xw2c3eXJSJLmZCzv2SRZDrwauLWVLkxyV5INSY5steOAR4Z229ZqM9VfBnyvqp7doy5JmjDdwybJS4DPAe+uqt3AFcDPAScBjwIfGkMP65JsSbJl586dvQ8nSdpD17BJcjCDoPlkVX0eoKoeq6ofVdWPgY8xOE0GsB04fmj3Za02U/1x4IgkB+1R/wlVdWVVraqqVUuXLp2fJydJGlnP2WgBPg7cX1UfHqofO7TZW4B72vJG4NwkhyY5AVgB3AbcDqxoM88OYTCJYGNVFXAzcE7bfy1wQ6/nI0nafz0vxPk64HeAu5Pc2WrvZTCb7CSggIeB3wWoqnuTXA/cx2Am2wVV9SOAJBcCNwJLgA1VdW97vPcA1yb5E+AbDMJNkjRhuoVNVX0NyDSrNs2yz6XApdPUN023X1U9xPOn4SRJE8orCEiSujNsJEndGTaSpO4MG0lSd4aNJKk7w0aS1J1hI0nqzrCRJHVn2EiSujNsJEndGTaSpO4MG0lSd4aNJKk7w0aS1J1hI0nqzrCRJHVn2EiSujNsJEndGTaSpO4MG0lSd4aNJKk7w0aS1J1hI0nqzrCRJHVn2EiSujNsJEndGTaSpO4MG0lSd4aNJKk7w0aS1J1hI0nqrlvYJDk+yc1J7ktyb5J3tfpRSTYnebD9PLLVk+TyJFuT3JVk5dBjrW3bP5hk7VD9NUnubvtcniS9no8kaf/1fGXzLPAHVXUicCpwQZITgfXATVW1Arip3Qc4A1jRbuuAK2AQTsDFwGuBU4CLpwKqbfOOof1Wd3w+kqT91C1squrRqvp6W34KuB84DlgDXN02uxo4uy2vAa6pgVuAI5IcC5wObK6qXVX1BLAZWN3WvbSqbqmqAq4ZeixJ0gQZy3s2SZYDrwZuBY6pqkfbqu8Cx7Tl44BHhnbb1mqz1bdNU5/u+OuSbEmyZefOnXN7MpKkfdY9bJK8BPgc8O6q2j28rr0iqd49VNWVVbWqqlYtXbq09+EkSXvoGjZJDmYQNJ+sqs+38mPtFBjt545W3w4cP7T7slabrb5smrokacL0nI0W4OPA/VX14aFVG4GpGWVrgRuG6ue1WWmnAk+20203AqclObJNDDgNuLGt253k1Has84YeS5I0QQ7q+NivA34HuDvJna32XuADwPVJzge+A7y1rdsEnAlsBX4AvB2gqnYleT9we9vufVW1qy2/E7gKeCHwpXaTJE2YbmFTVV8DZvrcy5um2b6AC2Z4rA3AhmnqW4BXzqFNSdIYeAUBSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSepupLBJ8su9G5EkLV6jvrL5aJLbkrwzyc907UiStOiMFDZV9avAbwHHA3ck+VSSN3ftTJK0aIz8nk1VPQj8EfAe4N8Alyf5VpJ/16s5SdLiMOp7Nq9KchlwP/BG4Deq6hfb8mUd+5MkLQIHjbjdnwN/Bby3qn44Vayqf0zyR106kyQtGqOGzVnAD6vqRwBJXgAcVlU/qKpPdOtOkrQojPqezZeBFw7df1GrSZK0V6OGzWFV9f2pO235RX1akiQtNqOGzT8nWTl1J8lrgB/Osr0kSc8ZNWzeDXwmyf9O8jXgOuDC2XZIsiHJjiT3DNUuSbI9yZ3tdubQuouSbE3yQJLTh+qrW21rkvVD9ROS3Nrq1yU5ZMTnIkkas1E/1Hk78Arg94D/BPxiVd2xl92uAlZPU7+sqk5qt00ASU4EzgV+qe3z0SRLkiwBPgKcAZwIvK1tC/DB9lg/DzwBnD/Kc5Ekjd++XIjzZOBVwEoGf/TPm23jqvoqsGvEx14DXFtVT1fVt4GtwCnttrWqHqqqZ4BrgTVJwuAzPp9t+18NnL0Pz0WSNEajfqjzE8B/B17PIHROBlbt5zEvTHJXO812ZKsdBzwytM22Vpup/jLge1X17B71mfpfl2RLki07d+7cz7YlSftr1M/ZrAJOrKqa4/GuAN4PVPv5IeA/zvEx96qqrgSuBFi1atVcn4MkaR+NehrtHuBfzPVgVfVYVf2oqn4MfIzBaTKA7Qwu8jllWavNVH8cOCLJQXvUJUkTaNSwORq4L8mNSTZO3fb1YEmOHbr7FgYhBrARODfJoUlOAFYAtwG3AyvazLNDGEwi2NheYd0MnNP2XwvcsK/9SJLGY9TTaJfs6wMn+TTwBuDoJNuAi4E3JDmJwWm0h4HfBaiqe5NcD9wHPAtcMHRpnAuBG4ElwIaqurcd4j3AtUn+BPgG8PF97VGSNB4jhU1V/X2SfwWsqKovJ3kRgz/+s+3ztmnKMwZCVV0KXDpNfROwaZr6Qzx/Gk6SNMFGnY32DgbTjP+ylY4D/qZTT5KkRWbU92wuAF4H7IbnvkjtZ3s1JUlaXEYNm6fbhyoBaLPAnEIsSRrJqGHz90neC7wwyZuBzwD/s19bkqTFZNSwWQ/sBO5mMINsE+A3dEqSRjLqbLSpD2F+rG87kqTFaKSwSfJtpnmPpqpePu8dSZIWnX25NtqUw4D/ABw1/+1IkhajUb/P5vGh2/aq+lPgrL6tSZIWi1FPo60cuvsCBq90Rn1VJEn6KTdqYHxoaPlZBtc1e+u8dyNJWpRGnY32670bkSQtXqOeRvv92dZX1Yfnpx1J0mK0L7PRTmbwvTMAv8Hg+2Ye7NGUJGlxGTVslgErq+opgCSXAF+sqt/u1ZgkafEY9XI1xwDPDN1/ptUkSdqrUV/ZXAPcluQL7f7ZwNVdOpIkLTqjzka7NMmXgF9tpbdX1Tf6tSVJWkxGPY0G8CJgd1X9GbAtyQmdepIkLTKjfi30xcB7gIta6WDgr3s1JUlaXEZ9ZfMW4DeBfwaoqn8EDu/VlCRpcRk1bJ6pqqJ9zUCSF/drSZK02IwaNtcn+UvgiCTvAL6MX6QmSRrRXmejJQlwHfAKYDfwC8AfV9Xmzr1JkhaJvYZNVVWSTVX1y4ABI0naZ6OeRvt6kpO7diJJWrRGvYLAa4HfTvIwgxlpYfCi51W9GpMkLR6zhk2Sf1lV/xc4fUz9SJIWob29svkbBld7/k6Sz1XVvx9DT5KkRWZv79lkaPnlPRuRJC1eewubmmFZkqSR7S1sfiXJ7iRPAa9qy7uTPJVk92w7JtmQZEeSe4ZqRyXZnOTB9vPIVk+Sy5NsTXJXkpVD+6xt2z+YZO1Q/TVJ7m77XN4+DyRJmkCzhk1VLamql1bV4VV1UFueuv/SvTz2VcDqPWrrgZuqagVwU7sPcAawot3WAVfAIJyAixnMhjsFuHgqoNo27xjab89jSZImxL58xcA+qaqvArv2KK/h+S9du5rBl7BN1a+pgVsYXBbnWAaz4DZX1a6qeoLBh0pXt3Uvrapb2jXbrhl6LEnShOkWNjM4pqoebcvf5fmvlj4OeGRou22tNlt92zT1aSVZl2RLki07d+6c2zOQJO2zcYfNc4avIj2GY11ZVauqatXSpUvHcUhJ0pBxh81j7RQY7eeOVt8OHD+03bJWm62+bJq6JGkCjTtsNgJTM8rWAjcM1c9rs9JOBZ5sp9tuBE5LcmSbGHAacGNbtzvJqW0W2nlDjyVJmjCjXhttnyX5NPAG4Ogk2xjMKvsAg+/GOR/4DvDWtvkm4ExgK/AD4O0AVbUryfuB29t276uqqUkH72Qw4+2FwJfaTZI0gbqFTVW9bYZVb5pm2wIumOFxNgAbpqlvAV45lx4lSeOxYBMEJEk/PQwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdbcgYZPk4SR3J7kzyZZWOyrJ5iQPtp9HtnqSXJ5ka5K7kqwcepy1bfsHk6xdiOciSdq7hXxl8+tVdVJVrWr31wM3VdUK4KZ2H+AMYEW7rQOugEE4ARcDrwVOAS6eCihJ0mSZpNNoa4Cr2/LVwNlD9Wtq4BbgiCTHAqcDm6tqV1U9AWwGVo+5Z0nSCBYqbAr4uyR3JFnXasdU1aNt+bvAMW35OOCRoX23tdpM9Z+QZF2SLUm27Ny5c76egyRpRAct0HFfX1Xbk/wssDnJt4ZXVlUlqfk6WFVdCVwJsGrVqnl7XEnSaBbklU1VbW8/dwBfYPCey2Pt9Bjt5462+Xbg+KHdl7XaTHVJ0oQZe9gkeXGSw6eWgdOAe4CNwNSMsrXADW15I3Bem5V2KvBkO912I3BakiPbxIDTWk2SNGEW4jTaMcAXkkwd/1NV9b+S3A5cn+R84DvAW9v2m4Azga3AD4C3A1TVriTvB25v272vqnaN72lIkkY19rCpqoeAX5mm/jjwpmnqBVwww2NtADbMd4+SpPk1SVOfJUmLlGEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLUnWEjSerOsJEkdWfYSJK6M2wkSd0ZNpKk7gwbSVJ3ho0kqTvDRpLU3UEL3YD2bvn6Ly7YsR/+wFkLdmxJi4evbCRJ3Rk2kqTuDBtJUneGjSSpO8NGktTdAR82SVYneSDJ1iTrF7ofSdJPOqCnPidZAnwEeDOwDbg9ycaqum9hO1s8FmratVOupcXlgA4b4BRga1U9BJDkWmANYNgc4PxskbS4HOhhcxzwyND9bcBr99woyTpgXbv7/SQP7Mexjgb+aT/2Gxf7m5vn+ssHF7iT6R0w4zeh7G/ufmEuOx/oYTOSqroSuHIuj5FkS1WtmqeW5p39zY39zY39zc2k9weDHuey/4E+QWA7cPzQ/WWtJkmaIAd62NwOrEhyQpJDgHOBjQvckyRpDwf0abSqejbJhcCNwBJgQ1Xd2+lwczoNNwb2Nzf2Nzf2NzeT3h/M9a2IqpqvRiRJmtaBfhpNknQAMGwkSd0ZNiOYxEviJHk4yd1J7pyakpjkqCSbkzzYfh45xn42JNmR5J6h2rT9ZODyNp53JVm5QP1dkmR7G8M7k5w5tO6i1t8DSU4fQ3/HJ7k5yX1J7k3yrlafiDGcpb+JGMMkhyW5Lck3W3//tdVPSHJr6+O6NpGIJIe2+1vb+uUL1N9VSb49NH4ntfrYf0facZck+UaSv23352/8qsrbLDcGEw/+D/By4BDgm8CJE9DXw8DRe9T+G7C+La8HPjjGfn4NWAncs7d+gDOBLwEBTgVuXaD+LgH+8zTbntj+nQ8FTmj//ks693cssLItHw78Q+tjIsZwlv4mYgzbOLykLR8M3NrG5Xrg3Fb/C+D32vI7gb9oy+cC13Uev5n6uwo4Z5rtx/470o77+8CngL9t9+dt/Hxls3fPXRKnqp4Bpi6JM4nWAFe35auBs8d14Kr6KrBrxH7WANfUwC3AEUmOXYD+ZrIGuLaqnq6qbwNbGfx30E1VPVpVX2/LTwH3M7hCxkSM4Sz9zWSsY9jG4fvt7sHtVsAbgc+2+p7jNzWunwXelCQL0N9Mxv47kmQZcBbwV+1+mMfxM2z2brpL4sz2SzYuBfxdkjsyuBwPwDFV9Whb/i5wzMK09pyZ+pmkMb2wnabYMHTacUH7a6ckXs3g/34nbgz36A8mZAzbKaA7gR3AZgavpr5XVc9O08Nz/bX1TwIvG2d/VTU1fpe28bssyaF79jdN7738KfBfgB+3+y9jHsfPsDlwvb6qVgJnABck+bXhlTV4fTsx89onrZ/mCuDngJOAR4EPLWg3QJKXAJ8D3l1Vu4fXTcIYTtPfxIxhVf2oqk5icCWRU4BXLFQv09mzvySvBC5i0OfJwFHAexaityT/FthRVXf0OoZhs3cTeUmcqtrefu4AvsDgl+uxqZfa7eeOhesQZulnIsa0qh5rfwB+DHyM50/zLEh/SQ5m8If8k1X1+VaemDGcrr9JG8PW0/eAm4F/zeD009SH14d7eK6/tv5ngMfH3N/qdnqyqupp4H+wcOP3OuA3kzzM4K2CNwJ/xjyOn2GzdxN3SZwkL05y+NQycBpwT+trbdtsLXDDwnT4nJn62Qic12bcnAo8OXSqaGz2OAf+FgZjONXfuW3GzQnACuC2zr0E+Dhwf1V9eGjVRIzhTP1NyhgmWZrkiLb8QgbfcXU/gz/q57TN9hy/qXE9B/hKe+U4zv6+NfQ/EmHwfsjw+I3t37eqLqqqZVW1nMHfuK9U1W8xn+PXe3bDYrgxmBnyDwzOAf/hBPTzcgYzfb4J3DvVE4NzpjcBDwJfBo4aY0+fZnAa5f8xOLd7/kz9MJhh85E2nncDqxaov0+049/VfnmOHdr+D1t/DwBnjKG/1zM4RXYXcGe7nTkpYzhLfxMxhsCrgG+0Pu4B/njod+U2BhMUPgMc2uqHtftb2/qXL1B/X2njdw/w1zw/Y23svyNDvb6B52ejzdv4ebkaSVJ3nkaTJHVn2EiSujNsJEndGTaSpO4MG0lSd4aNJKk7w0aS1N3/B3qaJs0rP2t5AAAAAElFTkSuQmCC\n",
      "text/plain": [
       "<Figure size 432x288 with 1 Axes>"
      ]
     },
     "metadata": {
      "needs_background": "light"
     },
     "output_type": "display_data"
    }
   ],
   "source": [
    "s_by_gene.plot.hist()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### aggregation functions\n",
    "Calculating the gene length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr th {\n",
       "        text-align: left;\n",
       "    }\n",
       "\n",
       "    .dataframe thead tr:last-of-type th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>End</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th></th>\n",
       "      <th>max</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genes</th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AT1G01010</th>\n",
       "      <td>5630</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01020</th>\n",
       "      <td>8666</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01030</th>\n",
       "      <td>12940</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01040</th>\n",
       "      <td>31079</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01050</th>\n",
       "      <td>32670</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             End\n",
       "             max\n",
       "Genes           \n",
       "AT1G01010   5630\n",
       "AT1G01020   8666\n",
       "AT1G01030  12940\n",
       "AT1G01040  31079\n",
       "AT1G01050  32670"
      ]
     },
     "execution_count": 62,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "s_by_gene_min = df1.groupby(\"Genes\").agg({'Start': ['min']})\n",
    "s_by_gene_max = df1.groupby(\"Genes\").agg({'End': ['max']})\n",
    "s_by_gene_max.head(5)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### The custom functions to calculate the gen length"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "metadata": {},
   "outputs": [],
   "source": [
    "##  This function calculated the gene length from a file previously filtered by genomic feature. \n",
    "##  custom function that returns a Series of all the aggregations. \n",
    "##  Use the Series index as labels for the new columns\n",
    "##        @x parameter is a dataframe containing the feature's lengths\n",
    "##  Return:\n",
    "##        @pd.series is a df with the calculated values into new columns\n",
    "\n",
    "def f(x):\n",
    "    d = {}\n",
    "    d['min'] = x['Start'].min()            #get the minimun value from each subgrouped set \n",
    "    d['max'] = x['End'].max()              #get the maximun value from each subgrouped set \n",
    "    d['gen_len'] = (d['max'] - d['min']) + 1      # difference btw data series \n",
    "    return pd.Series(d, index=['min', 'max', 'gen_len'])   #index as new column in the df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "metadata": {},
   "outputs": [],
   "source": [
    "## call the agg function\n",
    "df_gene_len = df1.groupby(\"Genes\").apply(f)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>gen_len</th>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>Genes</th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "      <th></th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AT1G01010</th>\n",
       "      <td>3760</td>\n",
       "      <td>5630</td>\n",
       "      <td>1871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01020</th>\n",
       "      <td>6915</td>\n",
       "      <td>8666</td>\n",
       "      <td>1752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01030</th>\n",
       "      <td>11864</td>\n",
       "      <td>12940</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01040</th>\n",
       "      <td>23519</td>\n",
       "      <td>31079</td>\n",
       "      <td>7561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AT1G01050</th>\n",
       "      <td>31382</td>\n",
       "      <td>32670</td>\n",
       "      <td>1289</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             min    max  gen_len\n",
       "Genes                           \n",
       "AT1G01010   3760   5630     1871\n",
       "AT1G01020   6915   8666     1752\n",
       "AT1G01030  11864  12940     1077\n",
       "AT1G01040  23519  31079     7561\n",
       "AT1G01050  31382  32670     1289"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_len.head(5)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Genes</th>\n",
       "      <th>min</th>\n",
       "      <th>max</th>\n",
       "      <th>gen_len</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>AT1G01010</td>\n",
       "      <td>3760</td>\n",
       "      <td>5630</td>\n",
       "      <td>1871</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>AT1G01020</td>\n",
       "      <td>6915</td>\n",
       "      <td>8666</td>\n",
       "      <td>1752</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>AT1G01030</td>\n",
       "      <td>11864</td>\n",
       "      <td>12940</td>\n",
       "      <td>1077</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>AT1G01040</td>\n",
       "      <td>23519</td>\n",
       "      <td>31079</td>\n",
       "      <td>7561</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>AT1G01050</td>\n",
       "      <td>31382</td>\n",
       "      <td>32670</td>\n",
       "      <td>1289</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27650</th>\n",
       "      <td>ATMG01350</td>\n",
       "      <td>346757</td>\n",
       "      <td>347194</td>\n",
       "      <td>438</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27651</th>\n",
       "      <td>ATMG01360</td>\n",
       "      <td>349830</td>\n",
       "      <td>351413</td>\n",
       "      <td>1584</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27652</th>\n",
       "      <td>ATMG01370</td>\n",
       "      <td>360717</td>\n",
       "      <td>361052</td>\n",
       "      <td>336</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27653</th>\n",
       "      <td>ATMG01400</td>\n",
       "      <td>363725</td>\n",
       "      <td>364042</td>\n",
       "      <td>318</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27654</th>\n",
       "      <td>ATMG01410</td>\n",
       "      <td>366086</td>\n",
       "      <td>366700</td>\n",
       "      <td>615</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>27655 rows × 4 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "           Genes     min     max  gen_len\n",
       "0      AT1G01010    3760    5630     1871\n",
       "1      AT1G01020    6915    8666     1752\n",
       "2      AT1G01030   11864   12940     1077\n",
       "3      AT1G01040   23519   31079     7561\n",
       "4      AT1G01050   31382   32670     1289\n",
       "...          ...     ...     ...      ...\n",
       "27650  ATMG01350  346757  347194      438\n",
       "27651  ATMG01360  349830  351413     1584\n",
       "27652  ATMG01370  360717  361052      336\n",
       "27653  ATMG01400  363725  364042      318\n",
       "27654  ATMG01410  366086  366700      615\n",
       "\n",
       "[27655 rows x 4 columns]"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "df_gene_len.reset_index()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 68,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "27655"
      ]
     },
     "execution_count": 68,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(df_gene_len)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "df_gene_len.to_csv('Araport11_GFF3_genes_lenghts.csv', sep='\\t', index=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "## Second and third options with agg functions\n",
    "## Aggregatin multiple functions to grouping by df\n",
    "\n",
    "grouped_genes = df1.groupby('ID').agg({'Start':'min', \n",
    "                                       'End':'max',\n",
    "                                       'len': lambda x: x.max() - x.min()})   \n",
    "## declaring with a function\n",
    "def max_min(x):\n",
    "    return x.max() - x.min()\n",
    "max_min.__name__ = 'Max minus Min'\n",
    "\n",
    "grouped_genes = df1.groupby('ID').agg({'Start':'min', \n",
    "                                       'End':'max',\n",
    "                                       'len': max_min})                   \n",
    "## Need to be fixed "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.6.9"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
